Initial commit for containerized Apache reverse proxy

brucehoff opened the pull request:
This PR defines an Apache2 reverse proxy suitable for use with RStudio or Jupyter notebooks. It includes GitHub workflows to run tests and build and publish the proxy as a Docker container to GHCR.

brucehoff committed:
migrating from Sge-Bionetworks org

brucehoff committed:
removed 'sudo' from ReadMe

xschildw commented on .github/workflows/docker_build.yml:
```diff
@@ -0,0 +1,48 @@
+          # Note we only publish containers on pushes to the main branch
+          push: |
+            ${{ github.event_name != 'pull_request' &&
+            github.ref == 'refs/heads/main'}}
```
Do we also push on tags? If so, I think in that case the github.ref is 'refs/tags/<tag_name>' and needs to be added to the push condition.
Q: should we only push on tag?

zaro0508 replied:
yes, we should build and publish the docker container on a tag.

zaro0508 commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,44 @@
+concurrency: ci-${{ github.ref }}
+
+jobs:
+  tests:
```
i still don't see a docker build step, e.g. `docker build -t [image_name] .`, in the tests job?  It should get triggered on a PR, a push to any branch and a tag.

xschildw replied:
I think it always builds and only pushes when the 'push' boolean is true. That would achieve the same as a build step in the test step would it not?

brucehoff replied:
@zaro0508 It's the `docker/build-push-action@v3.2.0` in `docker_build.yml` that builds and pushes the image.  Does that make sense?

zaro0508 replied:
> I think it always builds and only pushes when the 'push' boolean is true. That would achieve the same as a build step in the test step would it not?

it's not the same because docker build will not run as part of the tests job.  it would run in the docker-build-jupyter or docker-build-rstudio jobs which may potentially push as well which is confusing.  to simpfly things we should also add docker build to the tests job.

zaro0508 commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,44 @@
+    branches:
+      - '*'
+    tags:
+      - '*'
```
i suggest a regex for numbers, e.g. `tags: [ 'v[0-9]+\.[0-9]+\.[0-9]+' ]`

brucehoff replied:
OK, but we can add that constraint at the step that pushes the container to GHCR. I.e., we can run the workflow "liberally" but restrict when we push the (tagged) image.

zaro0508 replied:
hmm, what is the use case for pushing a tag that's not formatted in `'v[0-9]+\.[0-9]+\.[0-9]+'`?

zaro0508 commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,44 @@
+      - '*'
+  pull_request:
+    branches:
+      - 'main'
```
i would suggest `branches: [ '*' ]`  because we want to run the `tests` job when there is a PR against any branch.

brucehoff committed:
feedback from CR

brucehoff committed:
we can run the workflow liberally, but be very selective about when we push to GHCR

brucehoff committed:
fixed whitespace

zaro0508 commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,44 @@
+        run: python -m pytest tests/ -s -v
+
+
+  docker-build-jupyter:
```
this job name and `docker-build-rstudio` is confusing because this job is calling docker_build.yml which is doing a build and potentially a push, not just a build.  This job will execute when there is a PR, the workflow for a PR should only build, not potentially push as well.  

Again, i would suggest that you add a docker build only action to the tests job.  Rename this job to something like `docker-build-push-jypyter` and put a gate on this entire job to only trigger on pushes and tags.

zaro0508 commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+cryptography==3.4.8
+pyjwt
+requests
+boto3
+cachetools
```
it would probably be a good idea to pin all dependencies

zaro0508 commented on .github/workflows/docker_build.yml:
```diff
@@ -0,0 +1,49 @@
+          build-args: notebook_type=${{ inputs.NOTEBOOK_TYPE }}
+          # Note we only publish containers on pushes to the main branch
+          push: |
+            ${{ github.event_name != 'pull_request' &&
+            (github.ref == 'refs/heads/*' ||
+            github.ref == 'refs/tags/v[0-9]+\\.[0-9]+\\.[0-9]+' )}}
```
IMO it would be simpler and easier to understand if you move these conditions to main.yml

zaro0508 commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,34 @@
+repos:
```
i would recommend finding a pre-commit hook for docker at https://pre-commit.com/hooks.html  and adding that to this file.  something like `dockerfilelint` would be good.

brucehoff replied:
This is a great idea, thanks.

zaro0508 commented on .github/workflows/docker_build.yml:
```diff
@@ -0,0 +1,49 @@
+name: Create and publish a Docker image
```
the name of this one should be different than what's in main.yml

brucehoff replied:
Done.

zaro0508 reviewed (changes requested):
would you mind adding/updating this PR's title and description to reflect it's purpose?
![image](https://github.com/Sage-Bionetworks-IT/notebook-reverse-proxy/assets/3697686/df8ed50b-d41c-4747-b336-919e02018718)

Also why does the old repo still exists at https://github.com/Sage-Bionetworks/notebook-reverse-proxy ?

Here's an example of a workflow that incorporates my suggestions https://github.com/Sage-Bionetworks-IT/notebook-reverse-proxy/pull/2

brucehoff committed:
incorporate suggested CI workflow for building and publishing container

brucehoff committed:
updated names of workflow ymls

brucehoff commented:
> would you mind adding/updating this PR's title and description to reflect it's purpose? ![image](https://private-user-images.githubusercontent.com/3697686/295010327-df8ed50b-d41c-4747-b336-919e02018718.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ5MDI2NTgsIm5iZiI6MTcwNDkwMjM1OCwicGF0aCI6Ii8zNjk3Njg2LzI5NTAxMDMyNy1kZjhlZDUwYi1kNDFjLTQ3NDctYjMzNi05MTllMDIwMTg3MTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDExMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMTBUMTU1OTE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDIzODE5YjM4ZTU5YTk4OGY4MWM2YWYyYWJmNmE3MjQwMzc0YmE1NjNiZmUwZDMzMjdhZGE2ZjBjODJhYzBhMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ESLJtRowG-ns8M7T2N6SUOxICYp-MfJEUnOn11Yq9eM)

Done.


> Also why does the old repo still exists at https://github.com/Sage-Bionetworks/notebook-reverse-proxy ?

I just deleted it.

> Here's an example of a workflow that incorporates my suggestions #2

I incorporated your suggestions in to the PR.

brucehoff committed:
fixed whitespace

zaro0508 reviewed (changes requested):
the workflow looks good now however not all feedback has been address.  could you please take another look and address all the feedback?

brucehoff committed:
pinned Python dependencies; added dockerfile linter

zaro0508 commented on requirements.txt:
```diff
@@ -1,5 +1,5 @@
-requests
-boto3
-cachetools
+pyjwt==2.8.0
+requests==2.31.0
+boto3==1.34.15
+cachetools==5.3.2
```
it's typically better to be a little flexible when pinning versions.  for example, `pyjwt~=2.8.0` will get all *compatible* versions 2.8.0 and higher.  i would suggest doing that for all libs in requirements.txt and requirements-dev.txt

zaro0508 commented on Dockerfile:
```diff
@@ -1,12 +1,11 @@
+# Update the repository sources list and install Apache
+RUN apt-get update && \
+apt-get install -y --no-install-recommends git zip apache2 apache2-dev curl pip && apt-get clean && \
+rm -rf /var/lib/apt/lists/*
```
what is the purpose of removing the apt repository list?

brucehoff replied:
It's required by the [linter](https://github.com/pryorda/dockerfilelint-precommit-hooks) you recommended

> Dockerfile linter........................................................Failed
> - hook id: dockerfilelint
> - exit code: 3
> 
> File:   Dockerfile
> Issues: 3
> 
> Line 6: RUN apt-get update
> Issue  Category      Title                 Description
>     1  Optimization  apt-get update with   Use of apt-get update should be paired with rm -rf
>                      matching cache rm     /var/lib/apt/lists/* in the same layer.

zaro0508 replied:
i think it might make sense to run before `apt-get update` but not after because [the next execution of apt-get install will fail without preceding it with an apt-get update](https://askubuntu.com/a/179992)

brucehoff replied:
Sadly, the linter doesn't permit such a change:  Changing the line to:

> RUN rm -rf /var/lib/apt/lists/* && apt-get update && \
> apt-get install -y --no-install-recommends python3-dev git unzip apache2 apache2-dev curl pip && apt-get clean
> 

results in:

> Line 6: RUN rm -rf /var/lib/apt/lists/* && apt-get update && \
> Issue  Category      Title                 Description
>     1  Optimization  apt-get update with   Use of apt-get update should be paired with rm -rf
>                      matching cache rm     /var/lib/apt/lists/* in the same layer.

zaro0508 replied:
that pre-commit hook actually runs [the dockerfilelint linter](https://github.com/replicatedhq/dockerfilelint) which has options to allow you to disable some of the linting rules. i think we can probably disable the `apt-get_missing_rm` rule.

brucehoff replied:
Honestly, I'm not concerned about "the next execution of apt-get install" since there is only one execution in the Dockerfile. We list all the packages together in a single line for efficiency.

brucehoff replied:
... @zaro0508 I added your requested change.

brucehoff committed:
Docker linter requirements and pin Python dependency versions

brucehoff committed:
supporessed docker linting for 'missing rm' and including recommended dependencies (which breaks the product)

zaro0508 commented on .dockerfilelintrc:
```diff
@@ -0,0 +1,3 @@
+rules:
+  apt-get_missing_rm: off
+  apt-get_recommends: off
```
seems ok, but is there a reason for disabling `apt-get_recommends` as well?

brucehoff replied:
Yes, it breaks the product.  I spent half a day figuring out how to get the Python dev dependencies added and then found that a bunch of SSL cert' dependencies went missing. Rather than spending days sleuthing that second problem it' wa far easier simply to suppress the dockerfilelint finding.  The only downside is that the generated image might be a little bigger than it otherwise would, but that does not impact our product.

xschildw reviewed (approved):
Did not know about dockerfilelint...
